### DISCUSSION TITLE
RFC: Use of Azure Storage Emulator in restore (and/or backup) command testing

### DISCUSSION THREAD

**Discussion by wsmelton** [dataplat/dbatools#7661] [Ideas]
Is there any interest in utilizing Azure Storage Emulator for our test with the restore command. The restore command is the only one I can see us actually testing against Azure Storage (backup to URL).

Appveyor confirms they [do have the emulator installed](https://www.appveyor.com/docs/windows-images-software/) for us (or at least it should be). If not we can [always just install it ourselves](https://github.com/appveyor/ci/issues/1702). 

The setup is simply getting the emulator initialized and setup on the given server we are testing on for the given instance. This is [all documented](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-emulator#start-and-initialize-the-storage-emulator). I can see in the restore test we are testing on shared access keys which for the emulator is possible but would require us to just put the Az module for the Storage Accounts on the server so we can generate that key. The access key though is documented and known key so there is no security issues in including this in the scripts.

@potatoqualitee // @niphlod  // @Stuart-Moore

**Comment by niphlod** [comment]
confirmed, as we use the default image the software is onboarded. Why should we use it though if tests already use backup to url ?

**Comment by wsmelton** [comment]
They are backing up to someone's Azure subscription. I know they have failed once already because someone had to go in and purge the storage account because we exceeded the file limit for the blob container.

**Comment by niphlod** [comment]
ah ok, so with the storage emulator we could clean it up "better" ?

**Comment by wsmelton** [comment]
Yes, it would not require cleanup at all if each test run is a new build
server.

If we are using the same image or VM for multiple runs then we just need to
make sure storage space is not consumed.

**Comment by niphlod** [comment]
okay, got it! we do not persist storage between builds so it should be all fine and dandy

**Comment by potatoqualitee** [comment]
k im down

**Comment by wsmelton** [comment]
I'll see if I can work on this this week. I have the emulator on my dev machine so will see if I can get the test to run against it properly.

**Comment by potatoqualitee** [comment]
This sounds like a fun github action to setup -- it's now called `azurite` https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azurite
